### DISCUSSION TITLE
Alternate Stretching Method? (Midtone Transfer Function)

### DISCUSSION THREAD

**Discussion by AlaskanAstro** [aaronwmorris/indi-allsky#1481] [Ideas]
Hello, wondering about the possibility of adding an alternate stretching method than simply gamma. Namely a Midtone Transfer Function. The math behind it is relatively simple and can produce some very nice results. This is the "auto stretch" method used by default in PixInsight (and I think Siril). Here's some documentation on it here: https://pixinsight.com/doc/tools/AutoHistogram/AutoHistogram.html and here: https://pixinsight.com/doc/tools/HistogramTransformation/HistogramTransformation.html as well as here: https://siril.readthedocs.io/en/latest/processing/stretching.html

**Comment by aaronwmorris** [comment]
I will look into this.

This is the complete function from the Siril site:

```
xp = (original – shadows) / (highlights - shadows)

pixel = ((midtones - 1) * xp) / (((2 * midtones - 1) * xp) – midtones)
```

**Reply by alexhelms** [reply]
Also check out the pixinsight implementation -- https://gitlab.com/pixinsight/PCL/-/blob/master/include/pcl/HistogramTransformation.h#L388 and how it gets applied https://gitlab.com/pixinsight/PCL/-/blob/master/include/pcl/DisplayFunction.h#L327

**Comment by aaronwmorris** [comment]
Merged #1484 with the new Midtone Transfer Function.

Unfortunately, I had to change the way the stretching was initialized to support multiple options which will force people to re-enable stretching.

The new stretching selection and algorithm is usable in the FITS processor, so it is easy to test adjusting settings there.

**Comment by AlaskanAstro** [comment]
Nice, thanks. Looking forward to playing with this. Does that mean that you can adjust the parameters without restarting the INDI sky service?

**Reply by aaronwmorris** [reply]
The stretch setting for the image capture requires a Reload, but the FITS processor view can be changed dynamically without a restart.

**Reply by AlaskanAstro** [reply]
Is there documentation for the FITS view? I'm having trouble figuring out how to actually load an image into it. 

Edit: found the option to save FITS. Hopefully self-explanatory now.

**Reply by aaronwmorris** [reply]
Excellent, you found it.  I realize it is not straightforward, but I have not found a good way to represent that to the user.

When you have a FITS file captured, it will match the date of the image in the Image viewer view, there is a link at the bottom to process the individual FITS associated with that image.  When you navigate directly to FITS Processor, it just loads the last FITS file saved.

You can also now view the dark frames via the FITS processor in the Darks view.

**Reply by gigaklatch** [reply]
With this stretching method it seems like I am able to bring out some of the color from the FITS I saved from the Aurora a few weeks back. Is there a way to have it applied to all of the FITS from that night and having the result made into a new timelapse video?

**Reply by aaronwmorris** [reply]
Sorry, I was out on vacation and forgot to respond to this.

There is no functionality to do this, but I could write something that could do this.

**Reply by gigaklatch** [reply]
No worries!

Yes please, that would be much appreciated!  Should I start a new thread about that, instead of hijacking this topic?
